gitlab/pkg/apis/sources/v1alpha1/gitlabsource_validation.go

```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2020 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+
+	"knative.dev/pkg/apis"
+)
+
+// Validate GitLab source object fields
+func (g *GitLabSource) Validate(ctx context.Context) *apis.FieldError {
+	return g.Spec.Validate(ctx).ViaField("spec")
+}
+
+// Validate GitLab source Spec object fields
+func (gs *GitLabSourceSpec) Validate(ctx context.Context) *apis.FieldError {
+	var errs *apis.FieldError
+
+	// Validate sink
+	if gs.Sink == nil {
+		fe := apis.ErrMissingField("sink")
+		errs = errs.Also(fe)
+	} else if fe := gs.Sink.Validate(ctx); fe != nil {
+		errs = errs.Also(fe.ViaField("sink"))
+	}
```

Comment:
cc @n3wscott for wisdom around Source validation and Defaulting.

Reply:
There is a helper for sink:

```go
	// Validate sink
	errs = errs.Also(cs.Sink.Validate(ctx).ViaField("sink"))
```

Reply:
also, `spec.Sink` is required, so it should not be a pointer.

Reply:
you are using the helper 😆 sorry. But the question likely comes from the pointer type and having more logic here than needed. We can look into pushing some of the source spec validation into the ducktype too next release to help this.